### PR TITLE
LDAP: Refactor and add ldap_passwd module

### PR DESCRIPTION
**Pull request by KellerFuchs**
##### SUMMARY
- [x] Refactor `ldap_attr` and `ldap_entry`, pulling shared code, documentation and specs in `__init__`.
- [x] Add a `ldap_passwd` module.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
notification/ldap

##### ANSIBLE VERSION
```
ansible 2.4.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/user/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.14 (default, Sep 17 2017, 18:50:44) [GCC 7.2.0]
```

##### Additional information

I needed such a module while working on [LDAP-related automation](https://github.com/hashbang/admin-tools/pull/112).

### CODE REVIEW

File: lib/ansible/module_utils/ldap.py
Comment (Akasurde) [inline lib/ansible/module_utils/ldap.py]: I would prefer `ldap_argument_spec`

File: lib/ansible/modules/net_tools/ldap/ldap_attr.py
Comment (Akasurde) [inline lib/ansible/modules/net_tools/ldap/ldap_attr.py]: Any reason for this move?

File: lib/ansible/modules/net_tools/ldap/ldap_attr.py
Comment (Akasurde) [inline lib/ansible/modules/net_tools/ldap/ldap_attr.py]: Please import specific things rather than everything.

File: lib/ansible/modules/net_tools/ldap/ldap_entry.py
Comment (Akasurde) [inline lib/ansible/modules/net_tools/ldap/ldap_entry.py]: Please import specific things rather than everything.

File: lib/ansible/modules/net_tools/ldap/ldap_passwd.py
Comment (Akasurde) [inline lib/ansible/modules/net_tools/ldap/ldap_passwd.py:44]: You can create documentation fragment for common parameters. You can see pre-existing documentation fragment for [ipa](https://github.com/ansible/ansible/blob/devel/lib/ansible/utils/module_docs_fragments/ipa.py).

File: lib/ansible/modules/net_tools/ldap/ldap_passwd.py
Comment (Akasurde) [inline lib/ansible/modules/net_tools/ldap/ldap_passwd.py]: ```# Copyright: (c) 2017,```

File: lib/ansible/module_utils/ldap.py
Comment (Akasurde) [inline lib/ansible/module_utils/ldap.py]: module_utils code is BSD licensed and modules are GNU GPLv3 licensed. I think we need to get sign-off from everyone who contributed to this code snippet before moving it to module_utils. You can refer this [comment](https://github.com/ansible/ansible/pull/26613#issuecomment-318097595) for process.

**Review by Akasurde — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/net_tools/ldap/ldap_passwd.py
Comment (KellerFuchs) [reply lib/ansible/modules/net_tools/ldap/ldap_passwd.py:44]: Thanks for the pointer; I tried factoring out the documentation of the common parameters in an earlier version of the PR, but didn't know this should be done through `module_docs_fragments`.

**Comment by KellerFuchs** [comment]
@psagers @jtyr During that refactoring I moved code from your `ldap_*` modules to a shared helper in `module_utils`. However, this involves a change of license (from GPLv3 to BSD) for the code that moved to `modules_utils`.
I would like to have your permission to relicense the code that was moved to `ldap.py`.

**Comment by psagers** [comment]
Couldn't possibly care less. I pushed a change to the [original repository](https://bitbucket.org/psagers/ansible-ldap), re-releasing the code under CC0. I'll leave the repo up a little longer.

**Comment by KellerFuchs** [comment]
@psagers OK, thanks.

**Comment by KellerFuchs** [comment]
ready_for_review

**Comment by KellerFuchs** [comment]
@Akasurde I think I handled all your comments (pending OK from @jtyr for relicensing).

**Comment by KellerFuchs** [comment]
@jtyr Ping?  This PR is blocked on your OK for relicensing some of the code.

**Comment by jtyr** [comment]
I'm fine with change of the license but would prefer to change of `modules_utils` to GPLv3 instead of changing the modules to BSD ;o)

**Comment by KellerFuchs** [comment]
@jtyr Am I right in understanding this as an OK (albeit a facetious one) ?

If so, we only need @Akasurde to confirm that the changes they requested were done.

**Comment by KellerFuchs** [comment]
@Akasurde Ping?  This should be good to merge.

**Comment by Akasurde** [comment]
@KellerFuchs Could you please address ansibot reported issues ?

**Comment by KellerFuchs** [comment]
@Akasurde I was currently looking at that, and it should be fixed (we will see once the CI run); it took a while because the error message was somewhat confusing: I think the actual error was that there was both `type: bool` and `choices: [yes, no]` set in the documentation fragment.

That existed prior to the PR, and was fixed by @dagwieers in cdd21e2170b9a47d84624c8c2f792a66f2b0fe57, but I accidentally reintroduced it during the rebase.

File: lib/ansible/modules/net_tools/ldap/ldap_passwd.py
Comment (jtyr) [inline lib/ansible/modules/net_tools/ldap/ldap_passwd.py]: I would rather remove this exception here and put it around the `ldap.initialize()` and `self.connection.passwd_s()` above with a more specific error message.

File: lib/ansible/modules/net_tools/ldap/ldap_passwd.py
Comment (jtyr) [inline lib/ansible/modules/net_tools/ldap/ldap_passwd.py]: I would name the method `passwd_check`.

File: lib/ansible/modules/net_tools/ldap/ldap_passwd.py
Comment (jtyr) [inline lib/ansible/modules/net_tools/ldap/ldap_passwd.py]: I would name the method `passwd_set`.

File: lib/ansible/modules/net_tools/ldap/ldap_passwd.py
Comment (jtyr) [inline lib/ansible/modules/net_tools/ldap/ldap_passwd.py]: You can remove `type='str'` as that's the default type.

File: lib/ansible/utils/module_docs_fragments/ldap.py
Comment (jtyr) [inline lib/ansible/utils/module_docs_fragments/ldap.py]: This file consists of texts from previously copyrighted file. You should respect the original copyright.

File: lib/ansible/modules/net_tools/ldap/ldap_passwd.py
Comment (jtyr) [inline lib/ansible/modules/net_tools/ldap/ldap_passwd.py]: Break the line after the `(` to unify the style across the file.

File: lib/ansible/modules/net_tools/ldap/ldap_passwd.py
Comment (jtyr) [inline lib/ansible/modules/net_tools/ldap/ldap_passwd.py]: Break the line after the `(` to unify the style across the file.

File: lib/ansible/modules/net_tools/ldap/ldap_passwd.py
Comment (jtyr) [inline lib/ansible/modules/net_tools/ldap/ldap_passwd.py]: I would format this like this:

```python
ANSIBLE_METADATA = {
    'metadata_version': '1.1',
    'status': ['preview'],
    'supported_by': 'community'
}
```

File: lib/ansible/modules/net_tools/ldap/ldap_passwd.py
Comment (jtyr) [inline lib/ansible/modules/net_tools/ldap/ldap_passwd.py]: You don't have to quote this string.

File: lib/ansible/module_utils/ldap.py
Comment (jtyr) [inline lib/ansible/module_utils/ldap.py]: This file is missing copyright.

File: lib/ansible/module_utils/ldap.py
Comment (jtyr) [inline lib/ansible/module_utils/ldap.py]: Break the line after the `(` to unify the style across the file.

**Review by jtyr — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/net_tools/ldap/ldap_passwd.py
Comment (KellerFuchs) [reply lib/ansible/modules/net_tools/ldap/ldap_passwd.py]: This followed the conventions from the LDAP library, but we can change that.

File: lib/ansible/modules/net_tools/ldap/ldap_passwd.py
Comment (KellerFuchs) [reply lib/ansible/modules/net_tools/ldap/ldap_passwd.py]: Same as above.

File: lib/ansible/modules/net_tools/ldap/ldap_passwd.py
Comment (KellerFuchs) [reply lib/ansible/modules/net_tools/ldap/ldap_passwd.py]: I did so for consistency, but I can remove the quoting.

File: lib/ansible/module_utils/ldap.py
Comment (KellerFuchs) [inline lib/ansible/module_utils/ldap.py:12]: For some reason, Github is not dismissing the comment, but I added the copyright info.

**Comment by KellerFuchs** [comment]
@jtyr Comments handled, please ACK.

File: lib/ansible/utils/module_docs_fragments/ldap.py
Comment (jtyr) [inline lib/ansible/utils/module_docs_fragments/ldap.py]: The right order should be:

```
# Copyright 2017-2018 Keller Fuchs (@kellerfuchs) <kellerfuchs@hashbang.sh>
# Copyright 2016 Jiri Tyr <jiri.tyr@gmail.com>
# Copyright 2016 Peter Sagerson <psagers@ignorare.net>
```

File: lib/ansible/module_utils/ldap.py
Comment (jtyr) [inline lib/ansible/module_utils/ldap.py]: I would move this blank line above the `HAS_LDAP`.

File: lib/ansible/module_utils/ldap.py
Comment (jtyr) [inline lib/ansible/module_utils/ldap.py]: Please wrap the line after the `(`.

File: lib/ansible/modules/net_tools/ldap/ldap_passwd.py
Comment (jtyr) [inline lib/ansible/modules/net_tools/ldap/ldap_passwd.py]: I would keep the `HAS_LDAP` local to this package. Just set it to `True` after you import it bellow.

File: lib/ansible/modules/net_tools/ldap/ldap_passwd.py
Comment (jtyr) [inline lib/ansible/modules/net_tools/ldap/ldap_passwd.py]: What does mean the `u` in  `u_con`? What about to call it just `con`?

File: lib/ansible/modules/net_tools/ldap/ldap_passwd.py
Comment (jtyr) [inline lib/ansible/modules/net_tools/ldap/ldap_passwd.py]: I would remove the emoji here ;o)

**Review by jtyr — CHANGES_REQUESTED** [review]

File: lib/ansible/module_utils/ldap.py
Comment (KellerFuchs) [reply lib/ansible/module_utils/ldap.py]: What's the rationale for it?

File: lib/ansible/utils/module_docs_fragments/ldap.py
Comment (KellerFuchs) [reply lib/ansible/utils/module_docs_fragments/ldap.py]: This was literally copy-pasted from the original file.

**Comment by KellerFuchs** [comment]
@jtyr FYI, if those are a syntax convention that Ansible actually uses and enforces, there should be a linter running as part of the CI.  It's not a good use or your or my time to go back and forth about what essentially amounts to indentation changes and reshuffling copyright lines.

Moreover, it would make the requests feel far less arbitrary, esp. when they concern things that were literally copy-pasted from Ansible's codebase...

File: lib/ansible/modules/net_tools/ldap/ldap_passwd.py
Comment (robbyoconnor) [reply lib/ansible/modules/net_tools/ldap/ldap_passwd.py]: awww no fun whatsoever!

File: lib/ansible/module_utils/ldap.py
Comment (RyanSquared) [reply lib/ansible/module_utils/ldap.py]: A lot of Python code typically has a blank line after imports.

Comment (jtyr) [reply lib/ansible/module_utils/ldap.py]: Exactly.  The new line should be right after the import and not after the global variable.

File: lib/ansible/utils/module_docs_fragments/ldap.py
Comment (jtyr) [reply lib/ansible/utils/module_docs_fragments/ldap.py]: The original modules have this:

```yaml
# (c) 2016, Peter Sagerson <psagers@ignorare.net>
# (c) 2016, Jiri Tyr <jiri.tyr@gmail.com>
```

Peter is the original author, I'm the person who polished and published his module. You are another contributor, so you should be at the end of the list.

**Review by jtyr — CHANGES_REQUESTED** [review]

**Comment by jtyr** [comment]
I'm sorry for being so nit picking but your change is based on my work so it's in my interest to maintain the same quality of the code.

File: lib/ansible/utils/module_docs_fragments/ldap.py
Comment (KellerFuchs) [reply lib/ansible/utils/module_docs_fragments/ldap.py]: Thanks for clarifying, because it was utterly unclear to me that those lines should be in chronological order, rather than reverse chronological + alphabetical order.

Comment (KellerFuchs) [reply lib/ansible/utils/module_docs_fragments/ldap.py]: Eeeeeh, this is literally the opposite of [what you previously said](https://github.com/ansible/ansible/pull/33040#discussion_r182894470)

**Comment by KellerFuchs** [comment]
@jtyr My issue is not about you being nitpicky.

My issue is that you ignored this PR for months, and now that you deign pay attention to it, I am getting seemingly-arbitrary requests to change things, and when I comply with those, I get a contradictory request right afterwards.  Case in point:
- you provide a [specific order](https://github.com/ansible/ansible/pull/33040#discussion_r182894470) you want for the copyright lines;
- I comply with the request in 6549474da59428350cd7005ea4b830d7c99a4e0d, uniformizing that on all concerned files (incl. the ones I did not create);
- not even a week later [you request the exact opposite](https://github.com/ansible/ansible/pull/33040#discussion_r184024375).

If you truly feel that the code quality is improved by the changes, feel free to push commits doing exactly that to this branch; Github has an “allow edits from maintainers” feature for a reason.
In any case, I am entirely done wasting my time complying with contradictory, pointless changes; please let me know if you have actually-relevant review comments...

File: lib/ansible/utils/module_docs_fragments/ldap.py
Comment (RyanSquared) [reply lib/ansible/utils/module_docs_fragments/ldap.py]: > You are another contributor, so you should be at the end of the list.

And yet for your comment six days ago, @KellerFuchs was on top?

**Review by Akasurde — APPROVED** [review]

**Comment by Akasurde** [comment]
@KellerFuchs @jtyr I updated PR with required changes. Let me know if we agree on changes, so that we can merge changes.

File: lib/ansible/utils/module_docs_fragments/ldap.py
Comment (jtyr) [inline lib/ansible/utils/module_docs_fragments/ldap.py]: You should indent all the lists in this file by two more spaces to keep it the same like in all other files.

File: lib/ansible/module_utils/ldap.py
Comment (jtyr) [inline lib/ansible/module_utils/ldap.py]: This should look like this:

```yaml
# -*- coding: utf-8 -*-

# Copyright: (c) 2016, Peter Sagerson <psagers@ignorare.net>
# Copyright: (c) 2016, Jiri Tyr <jiri.tyr@gmail.com>
# Copyright: (c) 2017-2018 Keller Fuchs (@kellerfuchs) <kellerfuchs@hashbang.sh>
#
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```

The same in all other files.

File: lib/ansible/modules/net_tools/ldap/ldap_attr.py
Comment (jtyr) [inline lib/ansible/modules/net_tools/ldap/ldap_attr.py]: Please remove the `r` here as there is no reason to use byte string for any of the string blocks. The same in all other files.

File: lib/ansible/modules/net_tools/ldap/ldap_attr.py
Comment (jtyr) [inline lib/ansible/modules/net_tools/ldap/ldap_attr.py]: You should keep the `HAS_LDAP = True` here and remove the import for it. The same in all other files.

File: lib/ansible/module_utils/ldap.py
Comment (jtyr) [inline lib/ansible/module_utils/ldap.py:51]: I like this simplification but I believe that it should done somewhere in the core (suggestion for another PR) and not on per-module bases.

**Review by jtyr — CHANGES_REQUESTED** [review]

File: lib/ansible/module_utils/ldap.py
Comment (gundalow) [reply lib/ansible/module_utils/ldap.py:51]: Perhaps this can be done in a future PR?
That way there can be focused discussion about one topic in a single PR?

Comment (gundalow) [reply lib/ansible/module_utils/ldap.py:51]: ie this PR is fine, we can look at moving into Core at a later point

Comment (Akasurde) [reply lib/ansible/module_utils/ldap.py:51]: Agreed.

**Comment by gundalow** [comment]
@jtyr Thanks for the review comments so far, can you confirm you are happy with the structure and the only remaining technical issues are

1. `HAS_LDAP` handeling
2. `fail(self, msg, exn):`, - we will keep what we have in this PR and look at moving this out in a follow up PR

File: lib/ansible/modules/net_tools/ldap/ldap_attr.py
Comment (webknjaz) [reply lib/ansible/modules/net_tools/ldap/ldap_attr.py]: correction: It's not a byte string, it's a raw-string literal

**Comment by jtyr** [comment]
In general I'm happy with this PR. After the few details are ironed out, I'm happy to approve it.

**Review by jtyr — APPROVED** [review]

**Comment by Akasurde** [comment]
@KellerFuchs Thanks for the contribution. @jtyr @robbyoconnor @gundalow @webknjaz @RyanSquared Thanks for the reviews and comments.

**Comment by KellerFuchs** [comment]
@Akasurde You are welcome; thanks for the review and fixups.

**Pull request merged**
